### PR TITLE
Resolve different type in list creation

### PR DESCRIPTION
**Pull request by andyfengHKU**
This PR applies getMaxLogicalType logic to list creation. So we can now handle cases like `[1, 2.0]`.

The logic is as follows
```
// resolve a combined type for children in bindFunc
// cast each child towards the combined type
```
To do so, I have to 
- Introduce parameter types in FunctionBindData because this information is only available after bindFunc
- Apply casting after bindFunc

On the side, this PR also tries to remove `isVarLength` field from function. But it's not trivial to do so. I think fundamentally I still need some kind of indicator to tell me if a function accepts any number of parameter. So keep it in `Function` now and remove it from constructor (otherwise our constructor are getting messed up)

### CODE REVIEW

File: src/function/vector_array_functions.cpp [hunk 1]
```diff
@@ -14,17 +15,28 @@ namespace kuzu {
...
 }
 
 function_set ArrayValueFunction::getFunctionSet() {
     function_set result;
-    result.push_back(std::make_unique<ScalarFunction>(name,
-        std::vector<LogicalTypeID>{LogicalTypeID::ANY}, LogicalTypeID::ARRAY,
-        ListCreationFunction::execFunc, nullptr, ArrayValueBindFunc, true /*  isVarLength */));
+    auto function =
+        std::make_unique<ScalarFunction>(name, std::vector<LogicalTypeID>{LogicalTypeID::ANY},
+            LogicalTypeID::ARRAY, ListCreationFunction::execFunc, nullptr, ArrayValueBindFunc);
```
Comment (acquamarin) [inline src/function/vector_array_functions.cpp:37]: Add a comment for nullptr

**Review by acquamarin — APPROVED** [review]

File: test/test_files/tinysnb/function/list.test [hunk 2]
```diff
@@ -72,12 +72,18 @@
 False
 False
 True
+-STATEMENT RETURN [1, 2.0];
+---- 1
+[1.000000,2.000000]
+-STATEMENT RETURN [null, 3.0, 3];
+---- 1
+[,3.000000,3.000000]
 -STATEMENT RETURN [1, 'a'];
```
Comment (acquamarin) [inline test/test_files/tinysnb/function/list.test:81]: Duckdb will convert 1 to string in such case:
```
D select [1, 'a'];
┌─────────────────────────┐
│ main.list_value(1, 'a') │
│        varchar[]        │
├─────────────────────────┤
│ [1, a]                  │
└─────────────────────────┘
```

Comment (andyfengHKU) [reply test/test_files/tinysnb/function/list.test:81]: I'm pretty sure this is no longer the case in new version of Duckdb

**Comment by andyfengHKU** [comment]
Merged in #3330 instead.

**Pull request closed without merging**